eventauth/eventauth.go

```diff
@@ -362,3 +368,199 @@ func (e *eventAllower) commonChecks(event Event) error {
+		m.senderID == m.targetID &&
+		len(event.PrevEvents) == 1 {
+		// prev_events is a list of 2 element lists of event IDs and hashes.
+		if len(event.PrevEvents[0]) != 2 {
+			return errorf("unparsable prev event")
+		}
+		// Unmarshall the event ID string.
+		var prevEventID string
+		if err := json.Unmarshal(event.PrevEvents[0][0], &prevEventID); err != nil {
+			return errorf("unparsable prev event")
+		}
+		if prevEventID == m.create.eventID {
+			// If this is the room creator joining the room directly after the
+			// the create event, then allow.
+			return nil
+		}
+		// Otherwise fall back to the normal checks.
+	}
+
+	if m.newMember.Membership == "invite" && len(m.newMember.ThirdPartyInvite) != 0 {
+		// Special case third party invites
+		// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L393
+		panic(fmt.Errorf("ThirdPartyInvite not implemented"))
+	}
+
+	if m.targetID == m.senderID {
+		// If the state_key and the sender are the same then this is an attempt
+		// by a user to update their own membership.
+		return m.membershipAllowedSelf()
+	}
+	// Otherwise this is an attempt to modify the membership of somebody else.
+	return m.membershipAllowedOther()
+}
+
+// membershipAllowedSelf determines if the change made by the user to their own membership is allowed.
+func (m *membershipAllower) membershipAllowedSelf() error {
+	if m.newMember.Membership == "join" {
+		// A user that is not in the room is allowed to join if the room
+		// join rules are "public".
+		if m.oldMember.Membership == "leave" && m.joinRule.JoinRule == "public" {
+			return nil
+		}
+		// An invited user is allowed to join if the join rules are "public"
+		if m.oldMember.Membership == "invite" && m.joinRule.JoinRule == "public" {
+			return nil
+		}
+		// An invited user is allowed to join if the join rules are "invite"
+		if m.oldMember.Membership == "invite" && m.joinRule.JoinRule == "invite" {
+			return nil
+		}
+		// A joined user is allowed to update their join.
+		if m.oldMember.Membership == "join" {
+			return nil
+		}
+	}
+	if m.newMember.Membership == "leave" {
+		// A joined user is allowed to leave the room.
+		if m.oldMember.Membership == "join" {
+			return nil
+		}
+		// An invited user is allowed to reject an invite.
+		if m.oldMember.Membership == "invite" {
+			return nil
+		}
+	}
+	return m.membershipFailed()
+}
+
+// membershipAllowedSelf determines if the user is allowed to change the membership of another user.
+func (m *membershipAllower) membershipAllowedOther() error {
+	senderLevel := m.powerLevels.userLevel(m.senderID)
+	targetLevel := m.powerLevels.userLevel(m.targetID)
+
+	// You may only modify the membership of another user if you are in the room.
+	if m.senderMember.Membership == "join" {
+		if m.newMember.Membership == "ban" {
+			// A user may ban another user if their level is high enough
+			if senderLevel >= m.powerLevels.banLevel &&
+				senderLevel > targetLevel {
```

Comment:
Please cite the line in Synapse because it's non-obvious that:

> A user may ban another user if their level is high enough

should have the `senderLevel > targetLevel` check.